server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -428,12 +423,23 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
                                         // since we already have the name/uuid combo in the snapshots array
                                         uuid = parser.text();
                                     }
-                                    snapshotIds.add(snapshots.get(uuid));
+                                    if (uuid != null) {
+                                        SnapshotId snapshotId = snapshots.get(uuid);
+                                        if (snapshotId != null) {
+                                            snapshotIds.add(snapshotId);
+                                        } else {
+                                            // A snapshotted index references a snapshot which does not exist in
+                                            // the list of snapshots. This can happen when multiple clusters in
+                                            // different versions create or delete snapshot in the same repository.
+                                            throw new ElasticsearchParseException("Unknown snapshot uuid [" + uuid
```

Comment:
I wonder if we should extend the message to say that the repository has become corrupted. Otherwise it's not clear to the user what the problem is.